### PR TITLE
perf(Android): getConstants() in thread to parallelize load, speed up for most

### PR DESCRIPTION
**Pull request by mcuelenaere**
## Description

RNDeviceModule.getConstants() can take a lot of time (~300ms on a Galaxy S7), so try to run it as soon as possible in the background so that by the time it gets called, the constants have already been calculated and we can just return the map.

## Checklist

* [X] I have tested this on a device/simulator for each compatible OS
* [X] I added the documentation in `README.md`
* [ ] I mentioned this change in `CHANGELOG.md`

### CODE REVIEW

**Comment by noomorph** [comment]
What are your measurements and how did you measure them?

**Comment by mcuelenaere** [comment]
This was measured by adding a ReactMarker listener and printing out durations of each step, where we focus here only on the GET_CONSTANTS part of RNDeviceInfo.

Below are 3 measurements from before and after the patch, each starting from a cold environment (app was killed). As you can see, the GET_CONSTANTS duration gets reduced to almost nothing.

BEFORE:
```
   +325.000000 ms --                    GET_CONSTANTS [START]: RNDeviceInfo
   +229.000000 ms --                CONVERT_CONSTANTS [START]: RNDeviceInfo
     +6.000000 ms --                  CONVERT_CONSTANTS [END]: RNDeviceInfo
     +0.000000 ms --                      GET_CONSTANTS [END]: RNDeviceInfo

   +288.000000 ms --                    GET_CONSTANTS [START]: RNDeviceInfo
   +265.000000 ms --                CONVERT_CONSTANTS [START]: RNDeviceInfo
     +3.000000 ms --                  CONVERT_CONSTANTS [END]: RNDeviceInfo
     +0.000000 ms --                      GET_CONSTANTS [END]: RNDeviceInfo

   +286.000000 ms --                    GET_CONSTANTS [START]: RNDeviceInfo
   +262.000000 ms --                CONVERT_CONSTANTS [START]: RNDeviceInfo
     +4.000000 ms --                  CONVERT_CONSTANTS [END]: RNDeviceInfo
     +0.000000 ms --                      GET_CONSTANTS [END]: RNDeviceInfo
```

AFTER:
```
   +276.000000 ms --                    GET_CONSTANTS [START]: RNDeviceInfo
     +0.000000 ms --                CONVERT_CONSTANTS [START]: RNDeviceInfo
     +1.000000 ms --                  CONVERT_CONSTANTS [END]: RNDeviceInfo
     +0.000000 ms --                      GET_CONSTANTS [END]: RNDeviceInfo

   +382.000000 ms --                    GET_CONSTANTS [START]: RNDeviceInfo
     +0.000000 ms --                CONVERT_CONSTANTS [START]: RNDeviceInfo
     +1.000000 ms --                  CONVERT_CONSTANTS [END]: RNDeviceInfo
     +0.000000 ms --                      GET_CONSTANTS [END]: RNDeviceInfo

   +280.000000 ms --                    GET_CONSTANTS [START]: RNDeviceInfo
     +0.000000 ms --                CONVERT_CONSTANTS [START]: RNDeviceInfo
     +1.000000 ms --                  CONVERT_CONSTANTS [END]: RNDeviceInfo
     +0.000000 ms --                      GET_CONSTANTS [END]: RNDeviceInfo
```

Your mileage may vary, since this depends on how much time is in between the `RNDeviceModule` constructor being called and the `getConstants()` method being invoked (+ scheduling and thread synchronization overhead).

tl;dr: if you have a decent amount of RN modules being loaded in your app (and your device has multiple cores), this should have a positive effect.

**Comment by mikehardy** [comment]
just marking it as "needs-changes" because that's a tag I review periodically even when PRs go to sleep

**Comment by mcuelenaere** [comment]
@mikehardy the behaviour is optional now and disabled by default

**Comment by mikehardy** [comment]
How did I miss that? Let me re-read, because that sounds perfect...

**Comment by mikehardy** [comment]
Oh you just did it haha, thought I was losing my mind. This does sound perfect now, I will re-read and probably merge

Thanks for the contribution!

File: android/src/main/java/com/learnium/RNDeviceInfo/RNDeviceModule.java [hunk 1]
```diff
@@ -61,11 +63,28 @@
...
   DeviceType deviceType;
 
+  Map<String, Object> constants;
+  AsyncTask<Void, Void, Map<String, Object>> futureConstants;
+
   public RNDeviceModule(ReactApplicationContext reactContext) {
+    this(reactContext, false);
+  }
+
+  public RNDeviceModule(ReactApplicationContext reactContext, boolean loadConstantsAsynchronously) {
```
Comment (mikehardy) [inline android/src/main/java/com/learnium/RNDeviceInfo/RNDeviceModule.java]: This makes it optional, but now for this method to be invoke with async load, does it need some corresponding hook in RNDeviceInfo.java? Like RNDeviceInfo needs two constructors with a similar flag, it stores the state, then when [RNDeviceInfo.createNativeModules is called](https://github.com/mcuelenaere/react-native-device-info/blob/55dc03dc465786f745ee260532891121da18547f/android/src/main/java/com/learnium/RNDeviceInfo/RNDeviceInfo.java#L16) it sends the flag through? 

Otherwise I'm not sure how the new async style is invoked since the default (and only currently available) style from linking is to just new RNDeviceInfo and there are no other control flow entrypoints to toggle the behavior https://github.com/mcuelenaere/react-native-device-info/blob/55dc03dc46/example/android/app/src/main/java/com/example/MainApplication.java#L27

**Comment by mikehardy** [comment]
I pulled this locally and have it ready for testing - I think it needs one more tweak after I requested it be optional then it's probably good to go

File: android/src/main/java/com/learnium/RNDeviceInfo/RNDeviceModule.java (hunk 1 above)
Comment (mcuelenaere) [reply android/src/main/java/com/learnium/RNDeviceInfo/RNDeviceModule.java]: Oh, you're right. It's because locally I'm using Turbo Modules and thus creating the RNDeviceModule myself. I'll adapt `RNDeviceInfo` to also take a constructor argument.

**Comment by mcuelenaere** [comment]
@mikehardy `loadConstantsAsynchronously` is now propagated from `RNDeviceInfo` to `RNDeviceModule`

**Comment by mikehardy** [comment]
I'll repull and retest - super-speedy thanks!

**Comment by mikehardy** [comment]
https://github.com/react-native-community/react-native-device-info/releases/tag/v2.1.0 🚀

**Comment by ItsNoHax** [comment]
@mikehardy @mcuelenaere First of all thanks a lot to both of you for the fix. I would just like to point out that the change log and the README should reflect these changes. I'm still unclear on whether I should use true or false in the constructor.

**Comment by mikehardy** [comment]
@ItsNoHax the release notes do mention it? https://github.com/react-native-community/react-native-device-info/blob/master/CHANGELOG.md#210 (they say use `true`)

I suppose the README for install could be updated, it might help to explain that the constructor even has an optional boolean argument, and what it means. Happy to take a PR for that - you can edit it directly using the GitHub UI and propose some words that you think would help people?

**Comment by ItsNoHax** [comment]
Done, let me know what you think: https://github.com/react-native-community/react-native-device-info/pull/697

I left it under the main installation header as I was afraid people might miss it when linking automatically.

**Comment by zhao812** [comment]
expo how install react-native-device-info

**Comment by elyalvarado** [comment]
Hi guys, how will this work with RN>=0.60 auto loader?

**Comment by mikehardy** [comment]
@jslok to be honest I am not sure - but I am curious myself as I am just migrating my work project to RN60 now.

My current understanding is that to initialize a module with a non-default constructor (specifically: one that takes an argument) in the auto-linking world, you have to use a react-native-config.js to *disable* autolinking, then manually link it on the platform you disabled.

@thymikee is that correct? auto-linking only works with default constructors?

I initially merged this patch as a new option and did not set the deferred method to default but assuming my understanding of auto-linking is correct and it will only work with a default constructor, I'd accept a patch that reversed which style was default now that it has been released a while and shown to be working.

**Comment by thymikee** [comment]
@mikehardy disabling linking in `react-native.config.js` affects both autolinking and `link`. 

Passing a custom constructor is possible through [`android.packageInstance`](https://github.com/react-native-community/cli/blob/master/docs/dependencies.md#platformsandroidpackageinstance):

```js
module.exports = {
  dependencies: {
    'react-native-device-info': {
      platforms: {
        android: {
          packageInstance: 'new RNDeviceInfo(true)'
        }
      }
    }
  }
};
```

**Comment by mikehardy** [comment]
@thymikee I really appreciate both your responsiveness, and your spot-on response, thank you! I promise I'll be good at auto-linking soon :-) 

Okay then, @jslok I would happily take either a PR that made the parallel version default + a quick doc snippet on how to disable it (with the snippet above I think but reversed to false) or a doc enhancement that had this snippet showing how to enable it.

I think it's ready for default though, I believe more will benefit from it's being enabled than will be harmed so I would prefer just enabling by default

**Comment by Dror-Bar** [comment]
I don't have a file named react-native.config.js. This is unclear from the read me if I should create it and/or where should I create it.

**Comment by mikehardy** [comment]
@Dror-Bar if you wanted to propose a quick readme PR that made it clear to you that would help. Specifically addressed at the confusion: the file does not exist in the repository, but if you do *not* want the parallel initialization, you will need to create the file, and it should contain the stanza prescribed in the react-native-config.js in the readme

If you are okay with the new default / parallel initialization is okay for your project, you do not need to do anything

My decision to merge this and alter the default was based on the assumption it was reasonably tested (so would not break anyone) and that in general it was a performance increase for everyone or at least did not harm things. If this is not true for your project I would love to know why in order to see if my assumption was correct - and I'd apologize for regressing your project also, that definitely wasn't intended
